### PR TITLE
[PF-1247] Add log-based alerting triggers, make LoggingUtils public

### PR DESCRIPTION
**Pull request by zloery**
This change adds a new utility method in LoggingUtils to log a message at `ERROR` severity with the additional key-value pair `terraLogBasedAlert: true`. We can set up log-based alerts for this key, and callers from any service can then trigger environment-appropriate alerts from log entries as needed.

This also exposes the `jsonFromString` and `structuredLogData` utils as public. It looks like these were intended to be available to library users, but they're currently package-private.

### CODE REVIEW

File: src/main/java/bio/terra/common/logging/LoggingUtils.java [hunk 1]
```diff
@@ -27,6 +27,25 @@
...
+
+  /**
+   * Wrapper around logger.error() which also includes the alertObject defined above.
+   *
+   * <p>This allows callers to log at ERROR level and also trigger any environment-appropriate
+   * logging set up wherever their code is running. For logs which also need to include objects,
+   * callers can just include the alertObject() as an additional structured object to log.
+   */
+  public static void logAlert(Logger logger, String message) {
+    logger.error(message, alertObject());
```
Comment (jaycarlton) [inline src/main/java/bio/terra/common/logging/LoggingUtils.java:46]: You could log at SEVERE level here I think. I so rarely get to do that.

File: src/test/java/bio/terra/common/logging/LoggingTestController.java [hunk 2]
```diff
@@ -55,4 +55,18 @@ public void testStructuredLogging() throws JsonProcessingException {
     // The GSON should look like {"foo": {"bar": "baz"}}
     LOG.info("GSON object", jsonObject);
   }
+
+  @GetMapping("/testAlertLogging")
+  public void testAlertLogging() throws JsonProcessingException {
```
Comment (jaycarlton) [inline src/test/java/bio/terra/common/logging/LoggingTestController.java:60]: I would split your two test cases into separate endpoints and test methods.

File: src/test/java/bio/terra/common/logging/LoggingTest.java [hunk 3]
```diff
@@ -200,6 +201,34 @@ public void testStructuredLogging(CapturedOutput capturedOutput) {
...
+  @Test
+  public void testAlertLogging(CapturedOutput capturedOutput) {
+    ResponseEntity<String> response =
+        testRestTemplate.getForEntity("/testAlertLogging", String.class);
+    assertThat(response.getStatusCode().value()).isEqualTo(200);
+
+    String[] lines = capturedOutput.getAll().split("\n");
+    String event1 = null;
+    String event2 = null;
+    for (String line : lines) {
```
Comment (jaycarlton) [inline src/test/java/bio/terra/common/logging/LoggingTest.java]: This is hard to follow for me.

**Review by jaycarlton — APPROVED** [review]

File: src/test/java/bio/terra/common/logging/LoggingTest.java [hunk 4]
```diff
@@ -200,6 +201,34 @@ public void testStructuredLogging(CapturedOutput capturedOutput) {
     assertThat((String) readJson(event4, "$.foo.bar")).isEqualTo("baz");
   }
 
+  @Test
+  public void testAlertLogging(CapturedOutput capturedOutput) {
```
Comment (yuhuyoyo) [inline src/test/java/bio/terra/common/logging/LoggingTest.java:196]: have never seen test methods with parameters. til

File: src/test/java/bio/terra/common/logging/LoggingTest.java [hunk 5]
```diff
@@ -200,6 +201,34 @@ public void testStructuredLogging(CapturedOutput capturedOutput) {
...
+    ResponseEntity<String> response =
+        testRestTemplate.getForEntity("/testAlertLogging", String.class);
+    assertThat(response.getStatusCode().value()).isEqualTo(200);
+
+    String[] lines = capturedOutput.getAll().split("\n");
+    String event1 = null;
+    String event2 = null;
+    for (String line : lines) {
+      String message = readJson(line, "$.message");
+      if (message != null && message.contains("test alert message")) {
```
Comment (yuhuyoyo) [inline src/test/java/bio/terra/common/logging/LoggingTest.java]: we should extract a package visible filelds for the error messages to share between the two test classes.

File: src/test/java/bio/terra/common/logging/LoggingTest.java [hunk 6]
```diff
@@ -200,6 +201,34 @@ public void testStructuredLogging(CapturedOutput capturedOutput) {
...
+        event2 = line;
+      }
+    }
+
+    assertThat(event1).isNotNull();
+    assertThat((String) readJson(event1, "$.severity")).isEqualTo("ERROR");
+    assertTrue((Boolean) readJson(event1, "$." + LoggingUtils.ALERT_KEY));
+    assertThat(event2).isNotNull();
+    assertThat((String) readJson(event2, "$.severity")).isEqualTo("ERROR");
+    assertThat((String) readJson(event2, "$.pojo.name")).isEqualTo("asdf");
```
Comment (yuhuyoyo) [inline src/test/java/bio/terra/common/logging/LoggingTest.java]: same with these

File: src/test/java/bio/terra/common/logging/LoggingTest.java (hunk 4 above)
Comment (yuhuyoyo) [reply src/test/java/bio/terra/common/logging/LoggingTest.java:196]: but also I really wonder what is this testing? is it testing the loggingTestController or loggingUtils..

**Review by yuhuyoyo — APPROVED** [review]

File: src/main/java/bio/terra/common/logging/LoggingUtils.java (hunk 1 above)
Comment (zloery) [reply src/main/java/bio/terra/common/logging/LoggingUtils.java:46]: Unfortunately that's a `java.util.Logging` level, not SLF4J :(

File: src/test/java/bio/terra/common/logging/LoggingTestController.java (hunk 2 above)
Comment (zloery) [reply src/test/java/bio/terra/common/logging/LoggingTestController.java:60]: Done

File: src/test/java/bio/terra/common/logging/LoggingTest.java (hunk 3 above)
Comment (zloery) [reply src/test/java/bio/terra/common/logging/LoggingTest.java]: Split this into two tests, it should be a bit easier now

File: src/test/java/bio/terra/common/logging/LoggingTest.java (hunk 4 above)
Comment (zloery) [reply src/test/java/bio/terra/common/logging/LoggingTest.java:196]: It's a bit tricky because this is a logging library - we have a toy Spring service using the library (`loggingTestController`), and this test is verifying the messages the toy service logs.

File: src/test/java/bio/terra/common/logging/LoggingTest.java (hunk 5 above)
Comment (zloery) [reply src/test/java/bio/terra/common/logging/LoggingTest.java]: pulled this logic into `getLogContainingMessage` for all tests to use

File: src/test/java/bio/terra/common/logging/LoggingTest.java (hunk 6 above)
Comment (zloery) [reply src/test/java/bio/terra/common/logging/LoggingTest.java]: refactored `readJson` to read objects properly

File: src/test/java/bio/terra/common/logging/LoggingTest.java [hunk 7]
```diff
@@ -193,30 +183,92 @@ public void testStructuredLogging(CapturedOutput capturedOutput) {
...
+    String[] lines = capturedOutput.getAll().split("\n");
+    String logLine = getLogContainingMessage(lines, "test alert message");
+    assertThat(logLine).isNotNull();
+    assertThat((String) readJson(logLine, "$.severity")).isEqualTo("ERROR");
+    assertTrue((Boolean) readJson(logLine, "$." + LoggingUtils.ALERT_KEY));
+  }
+
+  @Test
+  public void testAlertLoggingWithObject(CapturedOutput capturedOutput) {
+    ResponseEntity<String> response =
```
Comment (jaycarlton) [inline src/test/java/bio/terra/common/logging/LoggingTest.java:209]: Much cleaner, thanks.

File: src/test/java/bio/terra/common/logging/LoggingTest.java [hunk 8]
```diff
@@ -193,30 +183,92 @@ public void testStructuredLogging(CapturedOutput capturedOutput) {
...
+  /**
+   * Uses the JsonPath library to extract data from a given path within a JSON string.
+   *
+   * @param line The line of text to extract from
+   * @param path The JSON object path to read
+   * @param clazz Class to return. If null, this is inferred by the JsonPath library.
+   * @return
+   */
+  @Nullable
+  private <T> T readJson(String line, String path, @Nullable Class<T> clazz) {
```
Comment (jaycarlton) [inline src/test/java/bio/terra/common/logging/LoggingTest.java:240]: That's cool. I didn't know how to do that.

**Review by jaycarlton — APPROVED** [review]

File: src/test/java/bio/terra/common/logging/LoggingTest.java (hunk 4 above)
Comment (yuhuyoyo) [reply src/test/java/bio/terra/common/logging/LoggingTest.java:196]: i see. thanks for explaining
